### ISSUE TITLE
Auto start after device boot? 

### ISSUE THREAD

**Issue by boyedarat** [futo-org/voice-input#17]
How do I get FUTO Voice Input to auto start after device boot?

I don't have Google TTS installed.
I have to open FUTO Voice Input at least once after boot, for any of my installed keyboards to detect it. If I don't do this, the keyboard app complains that there is no Voice Input installed. I have FUTO Voice Input enabled in Keyboard and Input settings.

So far it works well with SwiftKey, only after I open FUTO Voice Input at least once.

**Comment by LearningAsIGo71** [comment]
Same problem here... I just discovered that when my Pixel 7 running GrapheneOS (GOS) restarts FUTO Voice stops working in OpenBoard and AnySoftKeyboard and as you said I have to open FUTO Voice to get it to work again. However, FUTO Voice **does** work with the default GrapheneOS (GOS) keyboard upon phone restart without me having to do anything. 

**Here is the scenario after the phone restarts:**

**GOS Keyboard:** I open my SMS (or Signal or Joplin or whatever) and the microphone appears in the GOS keyboard and when I click it, FUTO Voice opens and works as expected. 

**OpenBoard:** The microphone icon is missing. With the OpenBoard selected as the keyboard in any app (i.e. SMS, Signal, etc.) I then open the FUTO Voice app and return to the previous app (i.e. SMS, Signal, etc.) and the microphone icon has appeared in the keyboard and FUTO Voice works as expected. 

**AnySoftKeyboard:** I tap the microphone icon and AnySoftKeyboard reports "Voice is not installed." With the AnySoftKeyboard selected as the keyboard in any app (i.e. SMS, Signal, etc.) I then open the FUTO Voice app and return to the previous app (i.e. SMS, Signal, etc.) and tap the microphone and FUTO Voice works. 

NOTES:

- I have to enable the FUTO Voice for both OpenBoard and AnySoftKeyboard following the instructions above. Getting FUTO Voice to work with one keyboard does not get it to work the other. Each keyboard needs to "fixed" separately as described above. 

- FUTO Voice is not available with any keyboard when operating within any browser's URL / address bar. With Vanadium, Brave, Firefox, Firefox Focus, and DuckDuckGo browsers, when I click in the address bar the keyboard pops up, but no microphone icon in any of the keyboards, including the GOS keyboard. This was never the case with Google's Keyboard with Speech Services. Keyboards with FUTO Voice within elements of webpages **do work** as expected, but not the address bar. This is a bit of an inconvenience as the most common way I search the Internet is from within the address bar.

**Comment by LearningAsIGo71** [comment]
A week has gone by and still no acknowledgement? I have my phone set to auto-restart everyday and constantly forget about his annoying bug until I go to speak some text into my phone and FUTO Voice is not available. This seems like it would be a quick and easy fix, but what do I know?

**Comment by abb128** [comment]
@LearningAsIGo71 Sorry for the lack of acknowledgment, I will look into this issue over the next few days

**Comment by abb128** [comment]
This issue seems to be caused by an Android 11+ feature that filters which packages are visible to apps to improve privacy - https://developer.android.com/training/package-visibility

AnySoftKeyboard/SwiftKey checks if voice input is installed by calling `PackageManager.queryIntentActivities` to get a list of packages which can handle voice input, and checking if this list is blank or not. Android may filter FUTO Voice Input out of this list, if it deems that including it may be a privacy issue. Rebooting your phone can be a consistent way to trigger this, and starting the FUTO Voice Input app can fix it.

Automatically starting the app on boot is not a consistent way to solve this and might not be possible in modern Android AFAIK. Instead, FUTO Voice Input needs to be specified in a `<queries>` block of the other app's manifest. AnySoftKeyboard already does this on the main branch by querying all IMEs, but the latest release (January 2022) that's currently on Play Store and F-Droid doesn't do this. SwiftKey queries for `RecognitionService` which isn't currently implemented.

I'll add a workaround for this in the next version so that it works with AnySoftKeyboard and SwiftKey without needing to start the app after reboot. I haven't been able to reproduce the issue with OpenBoard though.

Also, regarding browser search, most browsers (including Vanadium, Brave, and Firefox) have their own voice input icon that appears when you're typing in the address bar, you can use that to search if the keyboard lacks the key.

**Comment by LearningAsIGo71** [comment]
Thank you for addressing this bug. After you saying you have not been able to reproduce the problem with OpenBoard I decided to try it on a second GOS Pixel phone I have (a Pixel 4 XL I have) and to my surprise it works just fine (i.e. the microphone icon for FUTO Voice still appears with the keyboard) after the phone restarts. Then on the phone in which I experience problem (a Pixel 7) I installed OpenBoard and FUTO Voice on a secondary profile and surprisingly it does NOT experience the problem upon restarting that the primary profile does. I'm not sure what is going on.

**Comment by LearningAsIGo71** [comment]
Correction: Turns out the secondary profile is exhibiting the same behavior as the primary profile. Still looking as to what could be the culprit.

**Comment by abb128** [comment]
I'll be closing this as v1.2.5 includes the fix for AnySoftKeyboard, SwiftKey, and possibly other keyboards. If it's still not working after reboot until opening on some specific keyboard and you're on v1.2.5+ (version should be visible on the settings home menu), post a comment or open another issue.

**Comment by 1kevbot** [comment]
Hello. I'm having the same experience. Using CalyxOS on a Pixel 6 Pro with Open Board. 

Seems as though after the latest update to the app the microphone to use the voice input on the keyboard disappears from all apps every time the phone is restarted.

The fix I've found is to simply open the Futo app (no changes made in the app though, just simply opening it). The microphone reappears on all apps after doing that.

**Comment by abb128** [comment]
@1kevbot Can you let me know where you downloaded OpenBoard from, and which version of OpenBoard you have installed?

**Comment by 1kevbot** [comment]
@abb128 Yes, I downloaded it from f-droid, v1.4.5.

**Comment by Kevinf63** [comment]
> @1kevbot Can you let me know where you downloaded OpenBoard from, and which version of OpenBoard you have installed?

Apologies, I know this is closed but I'm willing to reopen or reference this issue in a new ticket to resolve. I'm on Graphene OS and a Pixel 7.

I have the same experience here. Logcat seems to be empty in terms of events when I check it. Attempting to force close Futo Voice Input does nothing.

Futo Voice Input was installed from the Google Play Store.
```
Version 1.2.7

org.futo.voiceinput
versionCode 20

targetSdk 33
minSdk 24

Installed: 18 Jan 2024; 09:05
```

Openboard was installed from F-Droid Repo.
```
Version 1.4.5

org.dslul.openboard.inputmethod.latin
versionCode 19

targetSdk 31
minSdk 19

Installed: 14 Dec 2023; 00:54
```

**Comment by LearningAsIGo71** [comment]
> > Apologies, I know this is closed but I'm willing to reopen or reference this issue in a new ticket to resolve. I'm on Graphene OS and a Pixel 7. > ```

Yep, this is still an issue for me. I have tried using AnySoftKeyboard, but find it less usable than the built-in GOS Keyboard. So I've just been using OpenBoard and opening FUTO Voice every time my phone restarts to get the mic to appear. Not ideal, but better the alternatives. 

One unusual aspect of my Pixel 7 is that I have a Work Profile setup. How about you, are you using a work profile?

**Comment by abb128** [comment]
@Kevinf63 @LearningAsIGo71 I'll move discussion of the issue with specifically OpenBoard to https://github.com/futo-org/voice-input/issues/49, I provided a response there
